### PR TITLE
aura-ext: limit the number of authored blocks per slot

### PR DESCRIPTION
**Pull request by slumber**
Resolves https://github.com/paritytech/cumulus/issues/2544

### CODE REVIEW

File: pallets/aura-ext/src/lib.rs [hunk 1]
```diff
@@ -68,7 +77,32 @@ pub mod pallet {
...
 
-			T::DbWeight::get().reads_writes(2, 1)
+			let new_slot = Aura::<T>::current_slot();
+			let velocity = T::ConsensusHook::block_processing_velocity();
+
+			let (new_slot, authored) = match SlotInfo::<T>::get() {
+				Some((slot, authored)) if slot == new_slot => {
+					// New slot is the same as previous one, increment the
+					// the counter if within the limit.
+					if authored < velocity.get() {
```
Comment (slumber) [inline pallets/aura-ext/src/lib.rs:87]: To match the issue description this should actually be
```rust
if authored <= velocity.get()
```

But it seems counter-intuitive to me. If there's a single core assigned to para, why would we allow `V + 1` blocks? This doesn't align with current behavior either.

@rphmeier

File: pallets/aura-ext/src/consensus_hook.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	fn from(value: NonZeroU32) -> Self {
+		Self(value)
+	}
+}
+
+/// The consensus hook for managing number of blocks authored per slot.
+pub trait ConsensusHook {
+	/// This hook is called on each block initialization to determine if it's valid to author
+	/// another block.
+	fn block_processing_velocity() -> BlockProcessingVelocity;
```
Comment (slumber) [inline pallets/aura-ext/src/consensus_hook.rs:45]: Is there any argument we could expect? Looks like the only option is some constant value, and if para takes another core they'd have to upgrade their runtime

File: pallets/aura-ext/src/lib.rs [hunk 3]
```diff
@@ -52,7 +55,13 @@ pub mod pallet {
 
 	/// The configuration trait.
 	#[pallet::config]
-	pub trait Config: pallet_aura::Config + frame_system::Config {}
+	pub trait Config: pallet_aura::Config + frame_system::Config {
+		/// Hook used to determine current velocity of block processing
+		/// by the relay chain.
+		///
+		/// Defines the limit of blocks authored per slot to `V`.
```
Comment (slumber) [inline pallets/aura-ext/src/lib.rs:62]: ditto here

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2809336

**Review by rphmeier — COMMENTED** [review]
I don't understand why this exposes a new trait `ConsensusHook`. 

Instead, we should be exporting one or more `struct`s from this pallet that can be plugged in as `pallet_parachain_system::ConsensusHook`s.

The only thing people can reasonably do right now is have a constant velocity which we clamp to '1', although in the future we will want to do stuff like read out the state of blockspace regions, claims queue, etc.

File: pallets/aura-ext/src/lib.rs [hunk 4]
```diff
@@ -68,7 +77,32 @@ pub mod pallet {
 			// Fetch the authorities once to get them into the storage proof of the PoV.
 			Authorities::<T>::get();
 
-			T::DbWeight::get().reads_writes(2, 1)
+			let new_slot = Aura::<T>::current_slot();
+			let velocity = T::ConsensusHook::block_processing_velocity();
+
+			let (new_slot, authored) = match SlotInfo::<T>::get() {
```
Comment (rphmeier) [inline pallets/aura-ext/src/lib.rs:83]: The only things `on_initialized` should be doing here is:
  1. ensure that the slot doesn't move backwards
  2. update the storage item.

The actual checks need to be done in the `pallet_parachain_system::ConsensusHook` implementation.

File: pallets/aura-ext/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply pallets/aura-ext/src/lib.rs:87]: Note that if this is implemented as a `pallet_parachain_system::ConsensusHook` with a fixed capacity, the rules we get allow authors to author N blocks, where N is the lesser of:
  * the amount of space remaining in the unincluded segment
  * V+1

There is more elaboration on the capacity/velocity duality in https://github.com/paritytech/cumulus/issues/2476 . But in effect, this is about building a backlog.

This _does_ align with current behavior as long as the unincluded segment capacity is 1. But this is a general piece of code for taking full advantage of asynchronous backing in Aura.

File: pallets/aura-ext/src/lib.rs (hunk 3 above)
Comment (rphmeier) [reply pallets/aura-ext/src/lib.rs:62]: Should be v+1, see #2476

**Comment by rphmeier** [comment]
I think we had a bit of an architectural mental model mismatch.

I do see where you're coming from, with introducing a separate `ConsensusHook` type for Aura that determines the velocity. That _could_ be combined with a `pallet_parachain_system::FixedCapacityConsensusHook` to get effectively the same behavior. At the moment, the determination of velocity and capacity are not very tightly linked.

But looking forward, both velocity and capacity will both need to be determined dynamically, e.g. by reading relay-parent state and seeing how many cores a parachain is allocated onto or similar things. That's why I chose the architecture I did: consensus engines should have one piece of code for determining capacity and velocity. `parachain-system` only needs to care about `Capacity`, not velocity. The consensus hook itself is responsible for defining velocity rules.

File: pallets/aura-ext/src/consensus_hook.rs (hunk 2 above)
Comment (rphmeier) [reply pallets/aura-ext/src/consensus_hook.rs:45]: Yeah. It'll have to be constant for now. This'll be a relatively easy amendment once we have completed the design of elastic scaling.

File: pallets/parachain-system/src/consensus_hook.rs [hunk 5]
```diff
@@ -100,3 +102,15 @@ impl<const N: u32> ConsensusHook for FixedCapacityUnincludedSegment<N> {
 ///
 /// This is a simple type alias around a fixed-capacity unincluded segment with a size of 1.
 pub type RequireParentIncluded = FixedCapacityUnincludedSegment<1>;
+
+impl<T: pallet_aura_ext::Config, const V: u32, const C: u32> ConsensusHook
+	for aura_consensus_hook::FixedVelocityConsensusHook<T, V, C>
+{
+	fn on_state_proof(_state_proof: &RelayChainStateProof) -> UnincludedSegmentCapacity {
+		Self::validate_slot();
```
Comment (slumber) [inline pallets/parachain-system/src/consensus_hook.rs:110]: This uses weight so we should probably modify consensus hook to return it

File: pallets/parachain-system/Cargo.toml [hunk 6]
```diff
@@ -34,6 +34,7 @@ xcm = { git = "https://github.com/paritytech/polkadot", default-features = false
 cumulus-pallet-parachain-system-proc-macro = { path = "proc-macro", default-features = false }
 cumulus-primitives-core = { path = "../../primitives/core", default-features = false }
 cumulus-primitives-parachain-inherent = { path = "../../primitives/parachain-inherent", default-features = false }
+cumulus-pallet-aura-ext = { path = "../aura-ext", default-features = false }
```
Comment (slumber) [inline pallets/parachain-system/Cargo.toml:37]: Should it be feature-gated?

File: pallets/parachain-system/src/consensus_hook.rs (hunk 5 above)
Comment (rphmeier) [reply pallets/parachain-system/src/consensus_hook.rs:110]: This implementation should be in the `pallet-aura-ext` crate. The dependency is backwards.

File: pallets/aura-ext/src/consensus_hook.rs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+use super::pallet;
+use frame_support::pallet_prelude::*;
+use sp_std::marker::PhantomData;
+
+/// A consensus hook for a fixed block processing velocity.
+pub struct FixedVelocityConsensusHook<T, const V: u32, const C: u32>(PhantomData<T>);
+
+impl<T: pallet::Config, const V: u32, const C: u32> FixedVelocityConsensusHook<T, V, C> {
+	/// Validates the number of authored blocks within the slot with respect to the `V + 1` limit.
+	pub fn validate_slot() -> Weight {
```
Comment (rphmeier) [inline pallets/aura-ext/src/consensus_hook.rs:31]: This implementation doesn't need to be public.

File: pallets/parachain-system/Cargo.toml (hunk 6 above)
Comment (rphmeier) [reply pallets/parachain-system/Cargo.toml:37]: Dependency is backwards. Aura-ext should depend on parachain-system and implement `ConsensusHook` in its own code.

This is a general principle in Rust development: expose traits in one crate and implement them in others.

File: pallets/aura-ext/src/consensus_hook.rs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+//! The definition of a [`FixedVelocityConsensusHook`] for consensus logic to manage
+//! block velocity.
+//!
+//! The velocity `V` refers to the rate of block processing by the relay chain.
+
+use super::pallet;
+use frame_support::pallet_prelude::*;
+use sp_std::marker::PhantomData;
+
+/// A consensus hook for a fixed block processing velocity.
```
Comment (rphmeier) [inline pallets/aura-ext/src/consensus_hook.rs:26]: docs: mention capacity, please.

File: pallets/aura-ext/src/consensus_hook.rs (hunk 7 above)
Comment (rphmeier) [reply pallets/aura-ext/src/consensus_hook.rs:31]: This function should just be the body of 

`fn on_state_proof`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2817212

File: pallets/parachain-system/src/lib.rs [hunk 9]
```diff
@@ -58,7 +58,7 @@ use sp_std::{cmp, collections::btree_map::BTreeMap, prelude::*};
 use xcm::latest::XcmHash;
 
 mod migration;
-mod relay_state_snapshot;
+pub mod relay_state_snapshot;
```
Comment (rphmeier) [inline pallets/parachain-system/src/lib.rs:61]: Either:
  * re-export the necessary types for this module and leave it private
  * add module docs to `relay_state_snapshot` and move this declaration alongside the other public modules.

**Review by rphmeier — APPROVED** [review]
Looks good modulo nit. Thanks.

**Pull request merged**